pmap/pmap.go

```diff
@@ -0,0 +1,272 @@
+package pmap
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+)
+
+const (
+	gfIanaPrivPortsStart = 49152
+	gfPortMax            = 65535
+)
+
+// PortType represents the type of state of the port
+type PortType int32
+
+// List of states an individual port can be in
+const (
+	GfPmapPortFree PortType = iota
+	GfPmapPortForeign
+	GfPmapPortLeased
+	GfPmapPortNone
+	GfPmapPortBrickserver
+)
+
+type portStatus struct {
+	Type       PortType
+	Bricknames []string // brick muxing
+	Xprt       interface{}
+}
+
+var registry = struct {
+	sync.RWMutex
+	BasePort  int
+	LastAlloc int
+	Ports     [gfPortMax + 1]portStatus
+}{}
+
+// NOTE: Export the functions defined here only when other parts of glusterd2
+//       actually starts using them.
+// TODO: There is no RDMA specific handling yet.
+
+func isPortFree(port int) bool {
+	conn, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		return false
+	}
+	conn.Close()
+	return true
+}
+
+func registrySearchByPort(port int) string {
```

Comment:
Yeah. I am not sure if this function is required. (even on GD1 side). If you still feel this is required, maybe returning a string[] instead would be better.